datahub-web-react/src/app/ingest/source/builder/RecipeForm/common.tsx

```diff
@@ -0,0 +1,420 @@
+import React from 'react';
+import { set, get } from 'lodash';
+
+export enum FieldType {
+    TEXT,
+    BOOLEAN,
+    LIST,
+    SELECT,
+}
+
+interface Option {
+    label: string;
+    value: string;
+}
+
+export interface RecipeField {
+    name: string;
+    label: string;
+    tooltip: string | React.ReactNode;
+    type: FieldType;
+    fieldPath: string | string[];
+    rules: any[] | null;
+    section?: string;
+    options?: Option[];
+    buttonLabel?: string;
+    getValueFromRecipeOverride?: (recipe: any) => any;
+    setValueOnRecipeOverride?: (recipe: any, value: any) => any;
+    placeholder?: string;
+}
+
+function clearFieldAndParents(recipe: any, fieldPath: string) {
+    set(recipe, fieldPath, undefined);
+
+    const updatedFieldPath = fieldPath.split('.').slice(0, -1).join('.'); // remove last item from fieldPath
+    if (updatedFieldPath) {
+        const parentKeys = Object.keys(get(recipe, updatedFieldPath));
+
+        // only child left is what we just set as undefined
+        if (parentKeys.length === 1) {
+            clearFieldAndParents(recipe, updatedFieldPath);
+        }
+    }
+    return recipe;
+}
+
+export function setFieldValueOnRecipe(recipe: any, value: any, fieldPath: string) {
+    const updatedRecipe = { ...recipe };
+    if (value !== undefined) {
+        if (value === null) {
+            clearFieldAndParents(updatedRecipe, fieldPath);
+            return updatedRecipe;
+        }
+        set(updatedRecipe, fieldPath, value);
+    }
+    return updatedRecipe;
+}
+
+export function setListValuesOnRecipe(recipe: any, values: string[] | undefined, fieldPath: string) {
+    const updatedRecipe = { ...recipe };
+    if (values !== undefined) {
+        const filteredValues: string[] | undefined = values.filter((v) => !!v);
+        return filteredValues.length
+            ? setFieldValueOnRecipe(updatedRecipe, filteredValues, fieldPath)
+            : setFieldValueOnRecipe(updatedRecipe, null, fieldPath);
+    }
+    return updatedRecipe;
+}
+
+export function setDottedFieldValuesOnRecipe(recipe: any, value: any, fieldPath: string[]) {
+    const updatedRecipe = { ...recipe };
+    if (value !== undefined) {
+        if (value === null) {
+            set(recipe, fieldPath, undefined);
+            clearFieldAndParents(updatedRecipe, fieldPath.slice(0, -1).join('.'));
+            return updatedRecipe;
+        }
+        set(updatedRecipe, fieldPath, value);
+    }
+    return updatedRecipe;
+}
```

Comment:
I would leave all of this in `utils.tsx` so that the utils file has all of our utility functions and types, but this common file simply has fields in it.